### DISCUSSION TITLE
Architecture testing: support for willcards

### DISCUSSION THREAD

**Discussion by clementbirkle** [pestphp/pest#1072] [Ideas]
Hi,

Currently, `Architecture` testing does not support willcards. If you are using an architecture per feature, it can be useful to employ wildcards.

For example:

```php
arch('app')
    ->expect('App\Modules\*\Models')
    ->toBeAbstract();
```

**Comment by localusercamp** [comment]
This will be very helpfull feature! But only if you separate files into folders _correctly_.

Also there will be nice to have more complex quering of entities. Something like this:
```php
// What is available for now
test('Models have no suffixes', function () {
    expect('App\Models')
        ->classes->not->toHaveSuffix('Model');
});

// What will be nice to have
test('Models have no suffixes', function () {
    expect('App\Models')
        ->targeting(function ($fileInfo) {
            // Check if file/class is exactly extends Model
        })->not->toHaveSuffix('Model')
        ->targeting(function ($fileInfo) {
            // Check if file/class is exactly extends Pivot
        })->toHaveSuffix('Pivot');
});

// This will be helpful if you targeting files with much more complex patterns 
// or if you filesystem structure does not separate entities in different folders 
// like Laravel `Model` and `Pivot` classes (both of them stored in `app/Models` folder by default)
```

@nunomaduro What do you think about that?

**Reply by Hipich** [reply]
yeap

**Comment by naxvog** [comment]
Got this working in a test environment as I need this functionality to effectively test in namespaces with multiple domains/modules. Need to see about adding unit tests for those cases as well.
I am fairly optimistic that I will be able to open a PR within a couple of days :)

edit: PR for non-globstar implementation: https://github.com/pestphp/pest-plugin-arch/pull/12
